tests/check/s3sink.c

```diff
@@ -147,6 +147,53 @@ GST_START_TEST (test_no_bucket_and_key_then_start_should_fail)
 }
 GST_END_TEST
 
+GST_START_TEST (test_location_property)
+{
+  GstElement *sink = gst_element_factory_make ("s3sink", "sink");
+  const gchar *bucket = "bucket";
+  const gchar *key = "key";
+  const gchar *location = "s3://bucket/key";
+
+  gchar *returned_bucket = NULL,
+    *returned_key = NULL,
+    *returned_location = NULL;
+
+  fail_if (sink == NULL);
+
+  // Verify setting bucket and key results in location == 's3://{bucket}/{key}'
+  g_object_set (sink,
+    "bucket", bucket,
+    "key", key,
+    NULL);
+  g_object_get (sink, "location", &returned_location, NULL);
+  fail_if (0 != g_ascii_strcasecmp(location, returned_location));
+  g_free (returned_location);
+  returned_location = NULL;
+
+  // Set the bucket and key to something different, then set location.
+  // Verify bucket and key match what is expected.
+  g_object_set (sink, "bucket", "", "key", "", NULL);
+  g_object_set (sink, "location", location, NULL);
```

Comment:
I'd like to know the interactions when location is set alongside bucket and key in the same `g_object_set()` call. E.g, does the outcome depend on the ordering of passing the parameters in `g_object_set()`? What if we were to call with the following:
```
 g_object_set (sink,
    "bucket", "bucket_1",
    "key", "key_1",
    "location", "s3://bucket_2/key_2"
    NULL);
```
What would happen and what should take precedence? If there is a defined interaction, we should add tests for these cases also.

Reply:
@louisharris any idea of what the defined behavior should be? Ideally, one should only use one or the other. Are there other examples of GStreamer elements that have properties like this?

Reply:
I think it makes sense if location takes precedence over bucket and key, if all 3 are supplied, bucket and key are ignored in favour of location parameter. An example of a gstreamer element that has some parameter dependencies is videotestsrc (https://gstreamer.freedesktop.org/documentation/videotestsrc/index.html?gi-language=c). Can see here the motion parameter depends on pattern=snow, else the parameter is ignored. The same logic here would be ideal, just to prevent any strange s3 path parsing being added due to different values supplied to bucket, key and location. Then we can add this rule to the README where the other parameter rules are.

Reply:
@louisharris my current thought is to add a full-fledged `location` property to the `GstS3UploaderConfig` struct and access it the same as everything else, then when creating the `MultipartUploader` we parse `bucket` and `key` from the `GstS3UploaderConfig::location` uri if it's set, otherwise we set directly from `GstS3UploaderConfig::bucket` and `GstS3UploaderConfig::key`. This would mean the element properties `bucket`, `key`, and `location` all exist by themselves and don't interact when setting/getting, but `location` is always used if it's been specified regardless of what `bucket` and `key` are.

Is that the behavior we're looking for?

Reply:
Yes I like this idea :)